src/charm.py

```diff
@@ -58,20 +93,60 @@ def _on_start(self, _) -> None:
     def _mysql(self):
         """Returns an instance of the MySQL object from mysqlsh_helpers."""
         if not self._mysql_helpers:
-            # TODO: replace stubbed arguments once mechanisms to generate them exist
-            # Mechanisms = generating user/pass and storing+retrieving them from peer databag.
+            mysql_configs = self._get_or_create_mysql_configs()
+
             self._mysql_helpers = MySQL(
-                "127.0.0.1",
-                "test_cluster",
-                "password",
+                mysql_configs["unit_ip"],
+                mysql_configs["cluster_name"],
+                mysql_configs["root_password"],
                 "serverconfig",
-                "serverconfigpassword",
+                mysql_configs["server_config_password"],
                 "clusteradmin",
-                "clusteradminpassword",
+                mysql_configs["cluster_admin_password"],
             )
 
         return self._mysql_helpers
 
+    def _get_or_create_mysql_configs(self):
+        peer_relation = self.model.get_relation(PEER)
+        if peer_relation is None:
+            raise Exception(f"Peer relation {PEER} has not yet been established")
+
+        mysql_configs = {
+            "unit_ip": self.model.get_binding(PEER).network.bind_address,
+            "cluster_name": peer_relation.data[self.app].get("cluster_name"),
+            "root_password": peer_relation.data[self.app].get("root_password"),
+            "server_config_password": peer_relation.data[self.app].get("server_config_password"),
+            "cluster_admin_password": peer_relation.data[self.app].get("cluster_admin_password"),
+        }
+
+        is_unit_leader = self.unit.is_leader()
+        has_missing_peer_data = any([value is None for value in mysql_configs.values()])
+        if has_missing_peer_data and not is_unit_leader:
```

Comment:
When raising this exception, it will not get threated in the hook code. Should'n this be mapped to new Exceptions and threated accordingly? (even though the hooks also test for leadership.)

Reply:
i was originally thinking that these exceptions would push the charm into an error state as these conditions should never occur. however, with the refactor (516bd9596e3be0e50d16a526485b3a330a60f2ff), we can make the assumption that the data will be available during the `start` event